Index: ui/history/transferhistoryitemdelegate.cpp
===================================================================
--- ui/history/transferhistoryitemdelegate.cpp	(revision 815911)
+++ ui/history/transferhistoryitemdelegate.cpp	(working copy)
@@ -43,7 +43,7 @@
     connect(m_actionDownload, SIGNAL(triggered()), SLOT(slotDownload()));
 
     m_actionDelete_Selected = new QAction(this);
-    m_actionDelete_Selected->setText(i18n("Delete selected"));
+    m_actionDelete_Selected->setText(i18nc("Delete selected history-item", "Delete selected"));
     m_actionDelete_Selected->setIcon(KIcon("edit-delete"));
     connect(m_actionDelete_Selected, SIGNAL(triggered()), SLOT(slotDeleteTransfer()));
 
Index: ui/history/transferhistory.cpp
===================================================================
--- ui/history/transferhistory.cpp	(revision 815911)
+++ ui/history/transferhistory.cpp	(working copy)
@@ -207,17 +207,17 @@
     switch (status)
     {
         case Job::Running:
-            return i18n("Running");
+            return i18nc("The transfer is running", "Running");
         case Job::Delayed:
-            return i18n("Delayed");
+            return i18nc("The transfer is delayed", "Delayed");
         case Job::Stopped:
-            return i18n("Stopped");
+            return i18nc("The transfer is stopped", "Stopped");
         case Job::Aborted:
-            return i18n("Aborted");
+            return i18nc("The transfer is aborted", "Aborted");
         case Job::Finished:
-            return i18n("Finished");
+            return i18nc("The transfer is finished", "Finished");
         default:
-            return "";
+            return QString();
     }
 }
 
