Index: core/plugin/transferfactory.cpp
===================================================================
--- core/plugin/transferfactory.cpp	(revision 746603)
+++ core/plugin/transferfactory.cpp	(working copy)
@@ -65,6 +65,7 @@
         popup->addSeparator();
 
     popup->addAction( KGet::actionCollection()->action("transfer_open_dest") );
+    popup->addAction( KGet::actionCollection()->action("transfer_open_file") );
     popup->addAction( KGet::actionCollection()->action("transfer_show_details") );
     popup->addAction( KGet::actionCollection()->action("transfer_copy_source_url") );
 
Index: mainwindow.cpp
===================================================================
--- mainwindow.cpp	(revision 746603)
+++ mainwindow.cpp	(working copy)
@@ -207,6 +207,11 @@
     openDestAction->setIcon(KIcon("document-open"));
     connect(openDestAction, SIGNAL(triggered()), SLOT(slotTransfersOpenDest()));
 
+    QAction *openFileAction = actionCollection()->addAction("transfer_open_file");
+    openFileAction->setText(i18n("Open File"));
+    openFileAction->setIcon(KIcon("document-open"));
+    connect(openFileAction, SIGNAL(triggered()), SLOT(slotTransfersOpenFile()));
+
     QAction *showDetailsAction = actionCollection()->addAction("transfer_show_details");
     showDetailsAction->setText(i18n("Show Details"));
     showDetailsAction->setIcon(KIcon("document-properties"));
@@ -452,6 +457,14 @@
     }
 }
 
+void MainWindow::slotTransfersOpenFile()
+{
+    foreach(TransferHandler * it, KGet::selectedTransfers())
+    {
+        new KRun(it->dest(), this, 0, true, false);
+    }
+}
+
 void MainWindow::slotTransfersShowDetails()
 {
     foreach(TransferHandler * it, KGet::selectedTransfers())
Index: mainwindow.h
===================================================================
--- mainwindow.h	(revision 746603)
+++ mainwindow.h	(working copy)
@@ -84,6 +84,7 @@
     void slotStartSelectedDownload();
     void slotDeleteSelected();
     void slotTransfersOpenDest();
+    void slotTransfersOpenFile();
     void slotTransfersShowDetails();
     void slotTransfersCopySourceUrl();
 
